tests/unit/test_pebble.py

```diff
@@ -180,3 +182,236 @@ def test_define_pebble_layer(
+            "environment": {"envVar": "value"},
+            "user": "ubuntu",
+            "user-id": 1000,
+            "group": "ubuntu",
+            "group-id": 1000,
+            "working-dir": "/tmp",
+            "on-success": "ignore",
+            "on-failure": "restart",
+            "on-check-failure": {"check": "restart"},
+            "backoff-delay": "10ms",
+            "backoff-factor": 1.2,
+            "backoff-limit": "1m",
+            "kill-delay": "5s",
+        }
+        _ = Service(**full_service)
+
+    def test_minimal_service(self):
+        _ = Service(override="merge", command="foo cmd")  # pyright: ignore
+
+    @pytest.mark.parametrize(
+        "bad_service,error",
+        [
+            # Missing fields
+            ({}, r"^2 validation errors[\s\S]*override[\s\S]*command"),
+            # Bad attributes values
+            (
+                {
+                    "override": "bad value",
+                    "command": "free text allowed",
+                    "startup": "bad value",
+                    "on-success": "bad value",
+                    "on-failure": "bad value",
+                    "on-check-failure": {"check": "bad value"},
+                },
+                r"^5 validation errors[\s\S]*"
+                r"override[\s\S]*unexpected value[\s\S]*'merge', 'replace'[\s\S]*"
+                r"startup[\s\S]*unexpected value[\s\S]*'enabled', 'disabled'[\s\S]*"
+                r"on-success[\s\S]*unexpected value[\s\S]*"
+                r"'restart', 'shutdown', 'ignore'[\s\S]*"
+                r"on-failure[\s\S]*unexpected value[\s\S]*"
+                r"'restart', 'shutdown', 'ignore'[\s\S]*"
+                r"on-check-failure[\s\S]*unexpected value[\s\S]*"
+                r"'restart', 'shutdown', 'ignore'[\s\S]*",
+            ),
+            # Bad attribute types
+            (
+                {
+                    "override": ["merge"],
+                    "command": ["not a string"],
+                    "summary": {"foo": "bar"},
+                    "after": "not a List",
+                    "environment": "not a Dict",
+                    "user-id": "not an int",
+                    "on-check-failure": {"check": ["not a Literal"]},
+                    "backoff-factor": "not a float",
+                },
+                r"^8 validation errors[\s\S]*"
+                r"unhashable type[\s\S]*"
+                r"str type expected[\s\S]*"
+                r"value is not a valid list[\s\S]*"
+                r"value is not a valid dict[\s\S]*"
+                r"value is not a valid integer[\s\S]*"
+                r"unhashable type[\s\S]*"
+                r"value is not a valid float[\s\S]*",
+            ),
+        ],
+    )
+    def test_bad_services(self, bad_service, error):
+        with pytest.raises(pydantic.ValidationError) as err:
+            _ = Service(**bad_service)
+        assert re.match(error, str(err.value)), "Unexpected Service validation error"
```

Comment:
I think `pytest.raises(pydantic.ValidationError, match=error)` will also work
edit: ah I see now that this test was moved

Reply:
nice, didn't know that. I've applied it to all similar tests in this file.